In [1]:
from decimal import *

### [Compute a (m times) modulo n efficiently](https://stackoverflow.com/a/41365271)

In [2]:
def geometric(a, k, n):
    """calculates (1 + a + a^2 + ... + a^(k - 1)) mod n)"""
    if k <= 2:
        return sum(a ** i for i in range(k)) % n
    else:
        m = k // 2
        b = pow(a, 2, n)
        g = ((1 + a) * geometric(b, m, n)) % n
        return g if k % 2 == 0 else (g + pow(a, k - 1, n)) % n

def modulo_repeat(a, m, n):
    """ returns aaaa...a (m times) modulo n"""
    k = len(str(a))
    r = pow(10, k, n)
    return (a * geometric(r, m, n)) % n

### Analyze pattern in values of $s(n)$
#### Clearly, we have $s(n) = (n\,(\text{mod}^*\,9))\times10^{\frac{n-1}{9}} + 10^{\frac{n-1}{9}} - 1$, where exponents of $10$ follow integer division

In [3]:
digit_sums = [sum(int(d) for d in str(n)) for n in range(1, 10 ** 6)]
s = {n: digit_sums.index(n) + 1 for n in range(1, 50)}
s

{1: 1,
 2: 2,
 3: 3,
 4: 4,
 5: 5,
 6: 6,
 7: 7,
 8: 8,
 9: 9,
 10: 19,
 11: 29,
 12: 39,
 13: 49,
 14: 59,
 15: 69,
 16: 79,
 17: 89,
 18: 99,
 19: 199,
 20: 299,
 21: 399,
 22: 499,
 23: 599,
 24: 699,
 25: 799,
 26: 899,
 27: 999,
 28: 1999,
 29: 2999,
 30: 3999,
 31: 4999,
 32: 5999,
 33: 6999,
 34: 7999,
 35: 8999,
 36: 9999,
 37: 19999,
 38: 29999,
 39: 39999,
 40: 49999,
 41: 59999,
 42: 69999,
 43: 79999,
 44: 89999,
 45: 99999,
 46: 199999,
 47: 299999,
 48: 399999,
 49: 499999}

In [4]:
{(n - 8, n): sum(s[k] for k in range(1, 50) if n - 9 < k <= n) for n in range(9, 50, 9)}

{(1, 9): 45, (10, 18): 531, (19, 27): 5391, (28, 36): 53991, (37, 45): 539991}

### Summing $s(n)$ inferred from above yields the following computation for $S(k)$
#### $S(k) = 45 + 531 + 5391 + 53991 + 539991 + 5399991 + \cdots + \sum_{n=9\lfloor\frac{k}{9}\rfloor + 1}^{k} s(n)$

Also note that the sum $531 + 5391 + 53991 + \cdots = 54(10+10^2+10^3+\cdots) - (9+9+9+\cdots)$

In [5]:
s = lambda n: ((lambda x: x if x != 0 else 9)(n % 9) + 1) * pow(10, (n - 1) // 9, 10 ** 9 + 7) - 1

def S(k):
    if k <= 9:
        # s(n) = n for 1 <= n <= 9
        return sum(n for n in range(1, k + 1))
    else:
        c_1 = k // 9
        return (45 + (540 * (0 if c_1 == 1 else modulo_repeat(1, c_1 - 1, 10 ** 9 + 7)) - 9 * (c_1 - 1)) + sum(s(n) for n in range(9 * c_1 + 1, k + 1))) % (10 ** 9 + 7)
    
S(20)

1074

### Definition of Fibonacci sequence numbers $\{f_n\}$ based on golden ratio computation

In [6]:
getcontext().prec = 30


def f(n):
    return int(round(Decimal(1) / Decimal(5).sqrt() * (((Decimal(1) + Decimal(5).sqrt()) / Decimal(2)) ** Decimal(n) - (
            (Decimal(1) - Decimal(5).sqrt()) / Decimal(2)) ** Decimal(n))))

### Final answer: $\sum_{i=2}^{90}S(f_i)\,(\text{mod }10^9+7)$

In [12]:
sum(S(f(i)) for i in range(2, 90 + 1)) % (10 ** 9 + 7)

922058210